`select_case_var` & `constexpr int` Cleanups Continued

dareumnam opened the pull request:
Pull request overview
---------------------
 - Remove `select_case_var`s and `constexpr int`s
   - DataHVACGlobals.hh
   - DataSizing.hh
   - General.cc
   - GeneralRoutines.cc
   - MixedAir.cc

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [X] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

dareumnam committed:
General.cc

dareumnam committed:
General cleanup

dareumnam committed:
GeneralRoutines.cc

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SCVandEnum

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SCVandEnum

dareumnam committed:
Mixedair erased an unused enum

dareumnam committed:
Mixedair erased an unused enum 2

dareumnam committed:
MixedAir 1

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SCVandEnum

dareumnam committed:
MixedAir 2

dareumnam committed:
MixedAir 3

dareumnam committed:
MixedAir 4

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SCVandEnum

dareumnam committed:
Address comments

dareumnam committed:
fix error

dareumnam committed:
fix error-erase overlaped enum

dareumnam committed:
Address comments

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SCVandEnum

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SCVandEnum

amirroth commented on src/EnergyPlus/DataSizing.hh:
```diff
@@ -65,17 +66,34 @@ struct EnergyPlusData;
+    };
 
-    extern Array1D_string const cOAFlowMethodTypes;
+    constexpr std::array<std::string_view, static_cast<int>(OAFlowCalcMethod::Num)> OAFlowCalcMethodNames{
```
Better to put this in the `.cc` file. If it is needed in more than one module, leave an `extern const` declaration here.

amirroth replied:
Please address this also.

amirroth commented on src/EnergyPlus/DataSizing.hh:
```diff
@@ -127,35 +145,28 @@ namespace DataSizing {
     constexpr int ZOAM_ProportionalControlDesOcc(9); // Use ASHRAE Standard 62.1-2004 or Trane Engineer's newsletter (volume 34-5)
     // to calculate the zone level outdoor air flow rates based on design occupancy
 
+    enum class SOAM
```
I would rather call this `SystemOAMethod`. I would not have guessed what `SOAM` is.

amirroth commented on src/EnergyPlus/General.cc:
```diff
@@ -91,6 +91,70 @@ using DataHVACGlobals::Bisection;
 // MODULE PARAMETER DEFINITIONS
 static constexpr std::string_view BlankString;
 
+enum class ReportType
```
Good to define these in the `.cc` file if they are only used within this module. Type scope reduction is also good. All scope reduction is good.

amirroth commented on src/EnergyPlus/General.cc:
```diff
@@ -1574,86 +1631,67 @@ void ScanForReports(EnergyPlusData &state,
-                }
+            AvailRpt CheckAvailRpt =
+                static_cast<AvailRpt>(getEnumerationValue(AvailRptNamesUC, UtilityRoutines::MakeUPPERCase(state.dataIPShortCut->cAlphaArgs(1))));
+            switch (CheckAvailRpt) {
```
It's generally better not to use control flow to set flags. Just set the flags to the control flow condition. For example:

```cpp
OutputEMSActuatorAvailSmall = (CheckAvailRpt == AvailRpt::NotByUniqueNames);
OutputEMSActuatorAvailFull = (CheckAvailRpt == AvailRpt::Verbose);
```

amirroth commented on src/EnergyPlus/General.cc:
```diff
@@ -1574,86 +1631,67 @@ void ScanForReports(EnergyPlusData &state,
-                    state.dataRuntimeLang->OutputEMSInternalVarsFull = false;
-                    state.dataRuntimeLang->OutputEMSInternalVarsSmall = false;
-                }
+            CheckAvailRpt =
```
Same.

amirroth commented on src/EnergyPlus/General.cc:
```diff
@@ -1574,86 +1631,67 @@ void ScanForReports(EnergyPlusData &state,
-                    state.dataRuntimeLang->OutputEMSErrors = false;
-                    state.dataRuntimeLang->OutputFullEMSTrace = false;
-                }
+            ERLdebugOutputLevel CheckERLlevel = static_cast<ERLdebugOutputLevel>(
```
Same.

amirroth commented on src/EnergyPlus/GeneralRoutines.cc:
```diff
@@ -113,6 +113,19 @@ enum GeneralRoutinesEquipNums
 
+enum class AirLoopHVACCompType
+{
+    Invalid = -1,
```
Probably want to CamelCase these.

amirroth replied:
Address.

amirroth commented on src/EnergyPlus/GeneralRoutines.cc:
```diff
@@ -1435,77 +1448,76 @@ void TestSupplyAirPathIntegrity(EnergyPlusData &state, bool &ErrFound)
+                        state.dataZoneEquip->SupplyAirPath(BCount).ComponentName(Count))
+                        continue;
+                    if (Count == 1 && AirPathNodeName != state.dataLoopNodes->NodeID(state.dataZonePlenum->ZoneSupPlenCond(Count2).InletNode)) {
+                        ShowSevereError(state, "Error in AirLoopHVAC:SupplyPath=" + state.dataZoneEquip->SupplyAirPath(BCount).Name);
```
`format`

amirroth commented on src/EnergyPlus/GeneralRoutines.cc:
```diff
@@ -1766,121 +1778,119 @@ void TestReturnAirPathIntegrity(EnergyPlusData &state, bool &ErrFound, Array2S_i
+
+            switch (CompType) {
+            case AirLoopHVACCompType::Zonemixer: {
+                for (Count2 = 1; Count2 <= state.dataMixerComponent->NumMixers; ++Count2) {
```
Scope reduction.

amirroth commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -1467,93 +1412,42 @@ void GetOAControllerInputs(EnergyPlusData &state)
+            if (thisVentilationMechanical.SystemOAMethod == DataSizing::SOAM::IAQPGC ||
+                thisVentilationMechanical.SystemOAMethod == DataSizing::SOAM::IAQPCOM) {
+                if (!state.dataContaminantBalance->Contaminant.GenericContamSimulation) {
+                    ShowSevereError(state,
```
`format`

amirroth commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -4125,29 +4023,34 @@ void VentilationMechanicalProps::CalcMechVentController(
+                        ZoneOABZ = max(ZoneOAPeople, ZoneOAArea, ZoneOAFlow, ZoneOAACH);
+                    } break;
+                    default: {
+                        ZoneOABZ = 0.0;
```
Is this a `0.0` or an `assert(false);`?

amirroth commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -4216,23 +4119,28 @@ void VentilationMechanicalProps::CalcMechVentController(
-                            } else {
-                                ZoneOABZ = 0.0;
-                            }
+                        switch (state.dataSize->OARequirements(OAIndex).OAFlowMethod) {
```
Feels strangely familiar.

amirroth replied:
Seeing this idiom repeatedly makes me think that `ZoneOAPeople`, `ZoneOAArea`, etc. should be replaced by `std::array<Real64, static_cast<int>(OAFlowCalcMethod::Num> ZoneOA;` and then just accessed as `ZoneOA[static_cast<int>OAFlowMethod];` instead of these ubiquitous `switch` statements.

dareumnam replied:
Thanks for your insight. Since there is another local variable named `ZoneOA`, I'll use the name of ` ZoneOACalc`

amirroth commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -5117,19 +5027,22 @@ void OAControllerProps::SizeOAController(EnergyPlusData &state)
-                        this->MaxOA = state.dataSize->FinalSysSizing(state.dataSize->CurSysNum).DesMainVolFlow;
-                    }
+                } break;
+                case DataHVACGlobals::AirDuctType::Other: {
```
`Main`, `Other`, and `default` can be combined into a single case, e.g.:

```cpp
case Main:
case Other:
default: {
...
} break;
```

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -775,15 +775,15 @@ namespace PhotovoltaicThermalCollectors {
                         DesignVolFlowRateDes = state.dataSize->FinalSysSizing(state.dataSize->CurSysNum).DesOutAirVolFlow;
```
`auto &thisFinalSysSizing`

amirroth commented on src/EnergyPlus/SimAirServingZones.cc:
```diff
@@ -1947,16 +1950,15 @@ void InitAirLoops(EnergyPlusData &state, bool const FirstHVACIteration) // TRUE
-                    if (CompTypeNum == CompType::WaterCoil_SimpleCool || CompTypeNum == CompType::WaterCoil_Cooling ||
-                        CompTypeNum == CompType::WaterCoil_DetailedCool || CompTypeNum == CompType::WaterCoil_CoolingHXAsst ||
-                        CompTypeNum == CompType::DXSystem) {
+                    if (CompTypeNum == CompType::WaterCoil_Cooling || CompTypeNum == CompType::WaterCoil_DetailedCool ||
```
`CompType`

amirroth commented on src/EnergyPlus/SimAirServingZones.hh:
```diff
@@ -105,11 +103,42 @@ namespace SimAirServingZones {
         Num
     };
 
+    constexpr std::array<std::string_view, static_cast<int>(SimAirServingZones::CompType::Num)> CompTypeNamesUC{
```
I like having the strings on the left rather than off-screeen. Better to move these to the `.cc` file.

amirroth replied:
Please move to `.cc` file, leaving a `extern const` declaration here if necessary.

dareumnam replied:
Since this array is used only in MixedAir.cc, I'll move this to MixedAir.cc.

amirroth commented on src/EnergyPlus/SizingManager.cc:
```diff
@@ -2452,26 +2453,26 @@ void ProcessInputOARequirements(EnergyPlusData &state,
 
     if (NumAlphas > 1) {
         if (UtilityRoutines::SameString(Alphas(2), "Flow/Person")) {
```
I guess `getEnumerationValue` is coming at some point?

amirroth commented on src/EnergyPlus/SizingManager.cc:
```diff
@@ -2492,38 +2493,44 @@ void ProcessInputOARequirements(EnergyPlusData &state,
-        state.dataSize->OARequirements(OAIndex).OAFlowMethod != ZOAM_ProportionalControlSchOcc &&
-        state.dataSize->OARequirements(OAIndex).OAFlowMethod != ZOAM_ProportionalControlDesOcc &&
-        state.dataSize->OARequirements(OAIndex).OAFlowMethod != ZOAM_IAQP) {
+    if (state.dataSize->OARequirements(OAIndex).OAFlowMethod != OAFlowCalcMethod::PerPerson &&
```
I would say `BITF` here, but I think we need to "fix" it first.

amirroth replied:
Actually no need to do `BITF` anymore. But `auto &thisOARequirements` could be helpful.

amirroth commented on tst/EnergyPlus/unit/UnitarySystem.unit.cc:
```diff
@@ -19124,7 +19124,7 @@ TEST_F(AirloopUnitarySysTest, WSHPVariableSpeedCoilSizing)
     state->dataSize->PlantSizData(1).PlantLoopName = "ColdWaterLoop";
     state->dataSize->PlantSizData(2).PlantLoopName = "HotWaterLoop";
-    state->dataSize->CurDuctType = DataHVACGlobals::Main;
+    state->dataSize->CurDuctType = DataHVACGlobals::AirDuctType::Main;
```
Not too many comments. You know what you're doing at this point. :+1:

dareumnam committed:
clang format

dareumnam committed:
address comments 1

amirroth commented on src/EnergyPlus/DataSizing.cc:
```diff
@@ -573,17 +563,17 @@ Real64 OARequirementsData::desFlowPerZonePerson(EnergyPlusData &state,
         // This is a DesignSpecification:OutdoorAir:SpaceList
         Real64 sumPeopleOA = 0.0;
         for (int dsoaCount = 1; dsoaCount <= this->numDSOA; ++dsoaCount) {
             auto const thisDSOA = state.dataSize->OARequirements(this->dsoaIndexes(dsoaCount));
```
You’re making a copy here. This needs to be `auto const &`.

amirroth replied:
Please address this.

amirroth commented on src/EnergyPlus/DataSizing.cc:
```diff
@@ -821,7 +805,7 @@ Real64 OARequirementsData::calcOAFlowRate(EnergyPlusData &state,
+                            if (this->OAFlowMethod == DataSizing::OAFlowCalcMethod::PCOccSch) {
                                 if (this->CO2MaxMinLimitErrorCount < 2) {
                                     ShowSevereError(state,
                                                     "CalcDesignSpecificationOutdoorAir DesignSpecification:OutdoorAir = \"" + this->Name + "\".");
```
`format`

amirroth commented on src/EnergyPlus/DataSizing.hh:
```diff
@@ -1236,7 +1248,7 @@ struct SizingData : BaseGlobalStruct
         this->CurZoneEqNum = 0;
```
I think you can just say `*this = SizingData();` or maybe (`new(this) SizingData();` would be better because it avoids the copy) for the body of `clear_state` because all of the defaults are in the declaration. See examples of this throughout the code, e.g., ZonePlenum.hh

amirroth commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -1995,23 +1889,23 @@ void GetOAControllerInputs(EnergyPlusData &state)
             }
 
-            if (state.dataMixedAir->VentilationMechanical(VentMechNum).SystemOAMethod == SOAM_ZoneSum) {
+            if (state.dataMixedAir->VentilationMechanical(VentMechNum).SystemOAMethod == DataSizing::SOAM::ZoneSum) {
```
This idiom can be implemented more simply using 

```cpp
static constexpr std::array<std::string_view, <static_cast<int>(SOAM::Num)> = {"ZoneSum,", "Standard62.1VentilationRateProcedure," ... };
```

amirroth commented on src/EnergyPlus/MixedAir.cc:
```diff
@@ -5117,19 +5027,22 @@ void OAControllerProps::SizeOAController(EnergyPlusData &state)
-                    auto const SELECT_CASE_var1(state.dataSize->CurDuctType);
-                    if (SELECT_CASE_var1 == Main) {
+                switch (state.dataSize->CurDuctType) {
+                case DataHVACGlobals::AirDuctType::Main: {
```
Same with `DesMainVolFlow`, `DesHeatingVolFlow`, etc. `std::array <Real64, static_cast<int>(AirDuctType::Cooling)> DesVolFlow;`

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
fix errors & address comments

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SCVandEnum

Myoldmopar commented:
@dareumnam I know you are working on steam stuff.  Let me know if you have any issues resolving the conflicts in the branch or dealing with the build failure(s).  I see they are just popping up on Linux, so I can reproduce/diagnose for you if needed.  Thanks!

dareumnam commented:
@Myoldmopar I'll resolve the conflict, push this branch, and check the CI result. I can address the rest of comments on another branch so that this branch could get merged in.

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
fix build error

dareumnam committed:
fix build error 2

dareumnam committed:
Revert "fix build error 2"

This reverts commit 6d1c90565e8a83f4161522698fa02c0d4d56b59d.

dareumnam committed:
fix build error 2

dareumnam committed:
clang format

dareumnam commented:
This branch needs a little more investigation about Linux-built failure, I'll dive into this right after I/O Freeze. 🙏

dareumnam committed:
pull devlop and merge conflicts

dareumnam committed:
fix error

dareumnam committed:
clang format

jmythms committed:
fix linux build errors

dareumnam committed:
resolve a warning

dareumnam committed:
pull develop and merge conflicts

dareumnam commented:
This is ready to review & get merged in.

amirroth commented on src/EnergyPlus/HVACManager.cc:
```diff
@@ -2022,7 +2022,7 @@ void ResolveAirLoopFlowLimits(EnergyPlusData &state)
 
     for (AirLoopIndex = 1; AirLoopIndex <= state.dataHVACGlobal->NumPrimaryAirSys; ++AirLoopIndex) {          // loop over the primary air loops
         for (SupplyIndex = 1; SupplyIndex <= AirToZoneNodeInfo(AirLoopIndex).NumSupplyNodes; ++SupplyIndex) { // loop over the air loop supply outlets
```
`auto &thisAirToZoneNodeInfo = state.dataAirLoop->AirToZoneNodeInfo(AirLoopIndex);` here would be more useful than `auto &AirToZoneNodeInfo` above.

dareumnam commented:
@amirroth I'm addressing the remaining comments on another branch to make this branch not too big. I'll also address the additional comments on that branch.

amirroth commented:
> @amirroth I'm addressing the remaining comments on another branch to make this branch not too big. I'll also address the additional comments on that branch.

Ok, thanks.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into SCVandEnum

Myoldmopar committed:
Clean up bad enumeration usage

Myoldmopar commented:
@dareumnam once you push those changes up I'll push up an additional change.  It's unrelated to your changes, just tagging a little fix onto here for convenience.

dareumnam committed:
address comments 2

dareumnam committed:
address comment

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into SCVandEnum

dareumnam commented:
I realized that there are only 2 commits which are not very many code lines to address the remaining comments, so I'll push the commits on this PR branch rather than another branch.

dareumnam committed:
fix custom check error

amirroth commented on src/EnergyPlus/SizingManager.cc:
```diff
@@ -2461,29 +2472,11 @@ void ProcessInputOARequirements(EnergyPlusData &state,
-                       Alphas(2), "PROPORTIONALCONTROLBASEDONDESIGNOCCUPANCY")) { // Proportional Control based on ASHRAE Standard 62.1-2004
-            state.dataSize->OARequirements(OAIndex).OAFlowMethod = DataSizing::OAFlowCalcMethod::PCDesOcc;
-        } else {
+        thisOARequirements = static_cast<OAFlowCalcMethod>(getEnumerationValue(OAFlowCalcMethodNamesUC, UtilityRoutines::MakeUPPERCase(Alphas(2))));
```
Ummm … shouldn’t this be thisOARequirements.OAMethod =?

dareumnam replied:
Oops, thanks so much for catching these!

dareumnam committed:
address comments 3

dareumnam committed:
address comments 4

dareumnam committed:
fix integration errors & address comments 5

dareumnam committed:
clang format

dareumnam commented:
All comments were addressed, I think this is ready for the final review & getting merged in.

dareumnam committed:
address one last comment that I missed

Myoldmopar committed:
Merge branch 'SCVandEnum' of github.com:NREL/EnergyPlus into SCVandEnum

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into SCVandEnum

Myoldmopar commented:
@dareumnam thanks for this, I attached an unrelated fix in there that shouldn't cause any issues (zero test failures or regressions locally).  If this is happy in the morning I'll merge it right in.

Myoldmopar commented:
This seems all ready to go, thanks @dareumnam  and thanks @amirroth for the comments.
